### PR TITLE
[ads] Periodically purge expired database rows

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/40265

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Verify that ad events (after 3 months for Rewards users and 30 days for non-Rewards users), ad history (after 30 days), creative set conversions (after the observation window), deposits (after 7 days), and transactions (after 3 months) are removed from the ads_service/database.sqlite database file. Additionally, ensure that the database is vacuumed 1 minute after the browser is launched and then once per day thereafter.

### CODE REVIEW

File: components/brave_ads/core/internal/ads_impl.h [hunk 1]
```diff
@@ -13,6 +13,7 @@
 #include "base/memory/weak_ptr.h"
 #include "brave/components/brave_ads/browser/ads_service_callback.h"
 #include "brave/components/brave_ads/core/internal/account/tokens/token_generator_interface.h"
+#include "brave/components/brave_ads/core/internal/database/database_maintenance.h"
```
Comment (aseren) [inline components/brave_ads/core/internal/ads_impl.h:16]: with the `std::unique_ptr<database::Maintenance>` it is enough to have forward declaration of database::Maintenance.

File: components/brave_ads/core/internal/ads_impl.h [hunk 2]
```diff
@@ -147,6 +143,10 @@ class AdsImpl final : public Ads {
   // state.
   GlobalState global_state_;
 
+  // Periodically handles database maintenance tasks, such as purging and
+  // vacuuming.
+  std::unique_ptr<database::Maintenance> database_maintenance_;
```
Comment (aseren) [inline components/brave_ads/core/internal/ads_impl.h:148]: we can possibly have this as  `database::Maintenance database_maintenance_;`.

File: components/brave_ads/core/internal/ads_impl.h (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/core/internal/ads_impl.h:16]: Fixed. And will follow up with transitioning database maintenance to AdsCore.

File: components/brave_ads/core/internal/ads_impl.h (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/core/internal/ads_impl.h]: Fixed. And will follow up with transitioning database maintenance to AdsCore.

**Review by aseren — APPROVED** [review]

**Pull request merged**